enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+
+## Motivation
+
+By design when OLM uninstalls an operator it does not remove any of the operator’s owned CustomResourceDefinitions (CRDs) and CRs in order to prevent data loss.
+The operator user or cluster admin is expected to manually delete CRDs, CRs and any other secondary resources created by the operator.
+While some secondary resources that have their [OwnerReference][owner-reference] set to a CR would be garbage collected, others such as off-cluster or cross-namespaced resources with no owner references would remain orphaned.
+
+Without the operator around to process deletion events of CRs, it becomes hard to cleanly remove deployed resources, especially when they exist outside of the cluster.
+
+### Goals
+
+- Clean up includes removal of CRs (operands) and CRDs
+- Operands and CRDs are not automatically removed
+- Upon operator uninstall a user can opt-in to the clean up behavior
+- The operand removal should have a timeout
+- After the operand removal timeout the operator should still be uninstalled
+- Exceeding the operand removal timeout should create an alert event
+
+### Non-Goals
+
+- OLM handles cleanup of anything other than CRs and CRDs
+- Handle custom cleanup logic like filtering/ordering the deletion of CRs 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator user I can instruct OLM to clean up all of my operator's CRs and CRDs when I uninstall my operator.
+
+#### Story 2
+
+As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Specify cleanup via Subscription
+
+To provide a way to opt-in for operand cleanup, the [`Subscription`][subscription] spec can include an optional field `spec.operandCleanup` which lets a user configure the cleanup behaviour.
+
+OLM would initiate cleanup upon the deletion of a `Subscription` instance, based on `spec.operandCleanup`:
+
+##### None (default)
+
+With `spec.operandCleanup.type: none` OLM will not clean up operands. This is also the default behavior if `spec.operandCleanup` is unspecified to make clean up opt-in.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: none
+```
+
+##### OLM managed
+
+With `operandCleanup.type: olm-managed` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces (`olm.targetNamespaces`) of that operator.
+Once the CRs have their finalizers cleared by the operator and are removed from the foreground, OLM will delete the CRD and then proceed with the removal of the operator.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: olm-managed
+    timeout: 30s
```

Comment:
I like the idea of a timeout / an alert for long running operations, but it's not clear to me how an admin (the one writing the description) will know what value to put here.

Cleaning up external services or vms could take a lot longer than sending a kill signal to an ephemeral process, for example - should there be a way for authors to indicate an expected time for cleanup?

Reply:
I think this is a good idea for the future. For the scope of this enhancement I would probably try to keep it as simple as possible to not take on too much complexity from the start.

Reply:
Yeah timeouts are not easy to configure (without some trial and error), but in the absence of a user-configurable timeout, we either:

1. Have no timeout
    - Allowing the operator removal to be blocked indefinitely in the event of a cleanup failure
2. Have a non-user-configurable timeout baked into OLM as a "sane" upper bound
    - This might be harder for us to set, where if it's too short could trigger cleanup-failed alert events for operators whose cleanup takes a long time.

I have a note on this in the risks section.

@dmesser WDYT? Simple as possible is no timeouts. Is that what you meant?

Reply:
I think it's good enough to have a sane upper bound as @dmesser suggested to get this going initially. I just wanted to point out the timeout should be atleast 30 seconds (the amount of time kubelet waits between SIGTERM and SIGKILL) in the case where some operand pod is unresponsive, as a base case. 

My thought is to bump the timeout to 15 minutes -- one entire OLM default sync cycle. If its not uninstalled by then, mark the cleanup a failure.

Reply:
Marking as resolved.
I've removed timeouts from the scope of this proposal per later comments.
https://github.com/operator-framework/enhancements/pull/46#discussion_r497841887